### ISSUE TITLE
`init;` as a potential `this.complete();` replacement

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#23005]
This issue is intended to capture the idea of using `init;` as a potential replacement for `this.complete();`, with the goal of addressing some of the concerns that have been raised about `this.complete();` in the past (https://github.com/Cray/chapel-private/issues/5239#issuecomment-1684185670, https://github.com/chapel-lang/chapel/issues/21335).  The idea behind this proposal is to treat `init;` as a special statement form recognized by the compiler and acted upon in the same way as `this.complete();` is today.

Pros:
* Uses a reserved word (to be), so enjoys syntax coloring
* Uses a pre-existing reserved word—and one associated with initializers—so doesn't require adding a new one
* Continues the trend of implementing all special methods using interfaces or reserved keywords (keywords here)
* Can be interpreted as (very tersely) saying "please finish initializing this" or "I assert that this is now initialized" (where the latter becomes less important if we were to adopt #22975)
  * There's a learning curve here, but arguably that's true with most of the approaches we've considered including `this.complete();`

Cons:
* Not a sibling to other single-statement forms in Chapel (`return;`, `break;`, `continue;`)
* Could be interpreted as creating a FCF reference to `init` and dropping it on the floor
  * but, we don't allow `init` to be called explicitly, so probably shouldn't (don't?) let FCFs be created for it anyway
  * `ident;` probably isn't usually interpreted by readers as "take a FCF to ident and drop it", esp. when highlighted as a keyword
  * we should potentially stop supporting this statement form anyway (arguably not interesting / an anti-pattern)—see #23000
* Could be confused with a `this.init()` call
  * But a single initializer may not have both a `this.init()` call and a `this.complete()` because the `this.init()` completely initializes the object and the compiler will error about the second; so these will not appear adjacent to one another
* Could be considered ugly/confusing next to a super.init call:
  ```chapel
  proc init(...args...) {
    super.init(...args...);
    init;
  }
  ```

**Comment by e-kayrakli** [comment]
I am not against the proposal, but I still find `this.complete();`/`complete();` to be a better alternative here. To me `complete()` is a method you call on your instance to initialize remaining fields that you haven't initialized. This operation is not such an intrinsic part of the language that warrants a keyword or some syntactic recipe.

I associate single-keyword statements like `init;` or `complete;` with control-flow rather than doing something on a value. A greedy solution to that problem is to also specify the thing that they operate on, e.g., `init this;` or `complete this;`. But that also has a problem of not being general enough as we can't do `init myVar;` or `complete x;`. So, we'll be putting two keywords together to create another "key phrase", which I don't think is right (though I always think of python's `yield from` which doesn't bother me too much).

To me a `complete` method that we provide by default in each type really captures the use of this feature well. It is just a convenience method to initialize fields that the user doesn't want to initialize manually. I do agree that we need to relax the compiler's rules around requirements for calling this method.

I find coming up with a more esoteric name (`this.initRemainingFields();`?) or turning this into a standalone function to be better solutions than coming up with some syntactic concoction to avoid potential clashes with user's custom `complete` method. 

---

I'll also note another potential misinterpretation of `init;`: this could be a call to a parenless initializer (which we don't have today, but should be a natural concept for a Chapel programmer)

**Comment by bradcray** [comment]
> This operation is not such an intrinsic part of the language that warrants a keyword or some syntactic recipe.

I'd agree with this statement more if we didn't require the call/feature to transition from the initialization phase of an initializer to the "use `this` phase" (#22975).  That is, as long as we can't write:

```chapel
record R {
  var x: int;
  proc init() {
    this.x = 42;
    writeln(this);
  }
}
```

this marker does feel like an intrinsic part of the language (to me) because we're not asking the compiler to initialize any fields for us; we're essentially using it to ask permission from the compiler/language to start using the object.

> I'll also note another potential misinterpretation of init;: this could be a call to a parenless initializer (which we don't have today, but should be a natural concept for a Chapel programmer)

Maybe we could pretend that's exactly what we're doing?  Calling a paren-less `init` procedure that's defined by the language and does exactly what your `this.complete()` does?  :)  Essentially, all types have a built-in, 0-argument, paren-less initializer that gets them ready for use and can only be invoked within initializers?

If we wanted to take that interpretation, it would suggest permitting `this.init;` as well.

**Comment by bradcray** [comment]
One thing I like about this proposal is that with `this.complete();`, I think Engin's "This is just a method, so I should be able to call it anywhere" perspective is difficult to argue against—even though I would because I strongly prefer the current behavior that it's only invokable within initializers and only once along any path through an initializer.  The reason I like this is that the capability, however it's spelled, is really only intended to be useful/used within initializers, and I think by making it more generally callable, we only make the potential for confusion stronger, particularly while we continue to have restrictions like "must be called equally along control flow paths within the initializer".  So keeping it look like a normal method call could make it seem surprising (even to someone as well-versed in the language as Engin) that it can't be called elsewhere assuming we retain the current rules.

Meanwhile, we already have a number of restrictions on how `init` can be called/used, where invoking it within initializers is the main (only?) place where it's permitted.  So leaning on `init;` feels like it allows us to keep the current strictness (can only call it once along any initializer path; must call it before using `this` within an initializer) without trying to rationalize as much why it can't be called outside of initializers, more than once, etc.

**Comment by bradcray** [comment]
I've split the question of whether, if we went with this approach, we should think of `init;` as being a shorthand for `this.init;` (where either one could be written) off into its own issue: #23021 as a sibling proposal.

**Comment by bradcray** [comment]
@DanilaFe implemented `init this;` as the replacement for `this.complete();` in https://github.com/chapel-lang/chapel/pull/23283, resolving this issue.
